core/src/uix/hooks/linter.clj

```diff
@@ -0,0 +1,157 @@
+    :loop #{loop}
+    :for  #{for doseq}
+    :iter-fn #{map mapv map-indexed filter filterv reduce reduce-kv keep keep-indexed
+               remove mapcat drop-while take-while group-by partition-by split-with
+               sort-by some}})
+
+(defmulti maybe-lint
+          (fn [[sym :as form]]
+            (reduce-kv
+              (fn [ret kw forms]
+                (if (forms sym)
+                  (reduced kw)
+                  ret))
+              form
+              forms)))
+
+(defmethod maybe-lint :default [form]
+  form)
+
+(defmethod maybe-lint :when [[_ test & body]]
+  (lint-hooks!* test :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) body))
+
+(defmethod maybe-lint :if [[_ test then else]]
+  (lint-hooks!* test :in-branch? false)
+  (lint-hooks!* then :in-branch? true)
+  (lint-hooks!* else :in-branch? true))
+
+(defmethod maybe-lint :logical [[_ test & tests]]
+  (lint-hooks!* test :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) tests))
+
+(defmethod maybe-lint :cond [[_ clause & clauses]]
+  (lint-hooks!* clause :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) clauses))
+
+(defmethod maybe-lint :condp [[_ pred e clause & clauses]]
+  (lint-hooks!* pred :in-branch? false)
+  (lint-hooks!* e :in-branch? false)
+  (lint-hooks!* clause :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) clauses))
+
+(defmethod maybe-lint :cond-threaded [[_ e & clauses]]
+  (lint-hooks!* e :in-branch? false)
+  (->> (partition 2 clauses)
+       (run! (fn [[test expr]]
+               (lint-hooks!* test :in-branch? false)
+               (lint-hooks!* expr :in-branch? true)))))
+
+(defmethod maybe-lint :some-threaded [[_ e clause & clauses]]
+  (lint-hooks!* e :in-branch? false)
+  (lint-hooks!* clause :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) clauses))
+
+(defmethod maybe-lint :case [[_ e clause & clauses]]
+  (lint-hooks!* e :in-branch? false)
+  (lint-hooks!* clause :in-branch? false)
+  (run! #(lint-hooks!* % :in-branch? true) clauses))
+
+(defmethod maybe-lint :loop [[_ bindings & body]]
+  (lint-hooks!* bindings :in-loop? false)
+  (run! #(lint-hooks!* % :in-loop? true) body))
+
+(defmethod maybe-lint :for [[_ bindings & body]]
+  (let [[binding & bindings] (partition 2 bindings)]
+    (lint-hooks!* (second binding) :in-loop? false)
+    (run! (fn [[v expr]] (lint-hooks!* expr :in-loop? true))
+          bindings)
+    (run! #(lint-hooks!* % :in-loop? true) body)))
+
+(defmethod maybe-lint :iter-fn [[_ f :as form]]
+  (when (and (list? f)
+             ('#{fn fn*} (first f))
+             (vector? (second f)))
+    (let [[_ _ & body] f]
+      (run! #(lint-hooks!* % :in-loop? true) body))))
+
+(defn add-error! [form type]
+  (swap! *component-context* update :errors conj {:source form
+                                                  :source-context *source-context*
+                                                  :type type}))
+
+(defn lint-hooks!*
```

Comment:
I am confused about the dynamic bindings. 🤔  How is this different from the following? What am I missing?

```clj
(defn lint-hooks!*
  [expr & {:keys [in-branch? in-loop?]}]
  (clojure.walk/prewalk
   (fn [form]
     (cond
       (and (list? form) (hook? (first form)))
       (do (when in-branch? (add-error! form ::hook-in-branch))
           (when in-loop? (add-error! form ::hook-in-loop))
           nil)

       (and (list? form) (or (not in-branch?) (not in-loop?)))
       (binding [*source-context* form]
         (maybe-lint form))

       :else form))
   expr)
  nil)
```

Reply:
Depending on the type of form being inspected in the current iteration it can be either in a loop, in a branch or in neither of them. When iterating through forms recursively `maybe-lint` fn calls `lint-hooks!*` and sets one of those boolean vars to `true` or `false` depending on detected conditions. Then the dynamic vars are rebinded to those new values, so that when inner forms are inspected we are still able to know if the current form is in a loop of a branch. I hope this makes sense.

Here's a very brief description flow of the algo
```clojure
(let [...] ;; in-branch? false, set initially
  (when ... ;; in-branch? false, no changes, reusing prev value
    (let [...] ;; in-branch? true, in a brach, updating to true
      (use-effect ...)))) ;; in-branch? true && hook => error
```

Reply:
Got it. 👍 
I would probably prefer to rather always pass the context along like so
```cljs
(defn lint-hooks!*
  [expr & {:keys [in-branch? in-loop?] :as ctx}]
  ;;; do stuff      
      (maybe-lint ctx form)
  ;;; more stuff
  nil)
 ```
 and then use `lint-hooks!*` somewhat like so
 ```cljs
 (lint-hooks!* expr (assoc ctx :in-branch? true))
 ```

which is arguably more verbose, but maybe less imperative... nevermind

Thank you for explaining it to me. 🙏